### ISSUE TITLE
Save Bar (Was: Pattern for combining Save, Cancel buttons)

### ISSUE THREAD

**Issue by warpech** [Starcounter/Guidelines#27]
Problem, explained in @andwah's words:

> If several apps expose a [Save] button (let's say on a "Person"). Both buttons may do different stuff on saving. (the end-user will probably "hide" one of the buttons.). Then one apps "Save" button will never be called and maybe important code will never be executed.

On the next App Architecture Task Force meeting, discuss the problem presented in https://github.com/Polyjuice/Polyjuice/issues/62 (compare it to Starcounter/Starcounter#2697).

Write new conclusions here and then close https://github.com/Polyjuice/Polyjuice/issues/62.

**Comment by warpech** [comment]
I think that the best idea presented in the issues above is: 

> I would like to see `Commiting`, `Commited`, `Restoring`, `Restored` events in transaction though.

[source](https://github.com/Polyjuice/Polyjuice/issues/62#issuecomment-112448999)

**Comment by warpech** [comment]
As it was discussed on the today's App Arch TF meeting, we should consider the following patterns:

1. `Save` button should only have `Transaction.Commit()`. `Cancel` button should only have `Transaction.Rollback()`. 
2. Any additional work can be done in a commit hook on a committed object, not directly in the `Input.Save` callback.
3. Check if there is any work performed in any app in the `Input.Cancel` callback.
4. In future, we will switch from having commit hooks to more general concept of transaction branches. @miyconst says that @mtornwall says that branches will support both scenarions:
   - object creation (execute on `new Person()`)
   - object saving (commit transaction)
5. Having a "Save" button directly in view is an anti-pattern. All our apps should have two levels of views:
   - lifecycle views (that have "Save", "Cancel", "Create new" buttons, nest blendable views)
   - blendable views (that **do not** have "Save", "Cancel", "Create new" buttons)
6. In future (Starcounter 2.4) we might consider transaction events such as: `Commiting`, `Commited`, `Restoring`, `Restored`.

The most important is point 5. We will work on creating a POC of it.

cc @tomalec

**Comment by warpech** [comment]
As it was discussed on the today's App Arch TF meeting, we should consider the following patterns:

1. `Save` button should only have `Transaction.Commit()`. `Cancel` button should only have `Transaction.Rollback()`. 
2. Any additional work can be done in a commit hook on a committed object, not directly in the `Input.Save` callback.
3. Check if there is any work performed in any app in the `Input.Cancel` callback.
4. In future, we will switch from having commit hooks to more general concept of transaction branches. @miyconst says that @mtornwall says that branches will support both scenarions:
   - object creation (execute on `new Person()`)
   - object saving (commit transaction)
5. Having a "Save" button directly in view is an anti-pattern. All our apps should have two levels of views:
   - lifecycle views (that have "Save", "Cancel", "Create new" buttons, nest blendable views)
   - blendable views (that **do not** have "Save", "Cancel", "Create new" buttons)
6. In future (Starcounter 2.4) we might consider transaction events such as: `Commiting`, `Commited`, `Restoring`, `Restored`.

The most important is point 5. We will work on creating a POC of it.

Something to add, @miyconst? cc @tomalec.

**Comment by warpech** [comment]
As reported by @bouguila in https://github.com/StarcounterApps/People/issues/93#issuecomment-310065662, the pattern of having only `Commit()` in `Save` handler is violated in multiple occasions in our apps.

For instance, UserAdmin does validation before committing: https://github.com/StarcounterApps/UserAdmin/blob/1fd8fde56d3e2e001b841419bea771a2e39038e5/src/UserAdmin/ViewModels/Administrator/SomebodyPage.json.cs#L102-L110

How can this be done better? We don't want this in `Save` handler and doing it in a commit hook is too late.

cc @miyconst, @chrhol

**Comment by tomalec** [comment]
> Having a "Save" button directly in view is an anti-pattern. All our apps should have two levels of views:
lifecycle views (that have "Save", "Cancel", "Create new" buttons, nest blendable views)
blendable views (that **do not** have "Save", "Cancel", "Create new" buttons)

I would say it will heavily limit designers freedom in blending.

**Comment by warpech** [comment]
Because you cannot put the "Save" button inside the page, only somewhere around the page - if I understand your concern correctly.

What do you suggest? To refrain from doing it until we have a better solution? This is the only feasible solution right now to me.

**Comment by tomalec** [comment]
> Because you cannot put the "Save" button inside the page, only somewhere around the page - if I understand your concern correctly.

Yes, not being able to blend "save" button somewhere in the middle of my custom rainbow-like grid layout.

I think that's the thing we present in many our demos, as this is typical, dead simply form rearranging that everybody understands.

Given that "Save button should only have `Transaction.Commit()`", then having multiple ones doing the same is not a big deal, as we can duck shoot them, with Shadow DOM.

If the application misbehaves and does something specific there, then does not act when duck shot, there's a reason to decrease the rating of that app. It's better to have a faulty UX due to app misbehavior, that from platform limitation.

**Comment by warpech** [comment]
> Given that "Save button should only have `Transaction.Commit()`", then having multiple ones doing the same is not a big deal, as we can duck shoot them, with Shadow DOM.

Should we duck shoot them automatically based on some hint, so you get a nice default?

@miyconst had some other arguments for having the "Save" button outside of the view. Let's hear what he has to add.

**Comment by tomalec** [comment]
>Should we duck shoot them automatically based on some hint, so you get a nice default?

For now, I would live defaults to be no-op and explicit result. We will see what would robot phase give us in terms of automation.

**Comment by miyconst** [comment]
>@miyconst had some other arguments for having the "Save" button outside of the view. Let's hear what he has to add.

My idea is that there are two kind of partials:

- Designed for blending from the very beginning. For example `Something's image`. This partial can be shown on `Something's` page only, and someone else has to provide that page. So, we can safely rely on that someone to provide the Save button.
- Designed for standalone usage, still can be blended. Such views are designed to work without any other views, and should include Save button.

The Save button can also be included in the first type of views, but there is no reason of doing so.

The standalone views should not abuse the Save button and use it only for committing the transaction.

**Comment by warpech** [comment]
@tomalec can you recall an example of a real world app that has the "Save" or "Cancel" button in the middle of a form?

If not, can you find such example within 1 hour of searching?

I can't, hence I think that the guideline presented in https://github.com/Starcounter/Guidelines/issues/27#issuecomment-309479099 will suffice for 90% of use cases. The whole point of a guideline is that you follow it unless you have a strong reason not to. In case you need to break the guideline in a specific case, that's fine.

**Comment by tomalec** [comment]
> @tomalec can you recall an example of a real world app that has the "Save" or "Cancel" button in the middle of a form?

We are not talking about just re-ordering, but also re-composing:

```html
<slot name="People/firstname"></slot>
<slot name="Images/avatar"></slot>
<paper-button>Submit</paper-buton>
<paper-dialog>
   <slot name="People/save"></slot>
   <paper-buton>Cancel</paper-button>
</paper-dialog>
```

Do you want me to find real-word example of the case above?

**Comment by tomalec** [comment]
Or even just:
```html
   <input slot="People/firstname">
   <button slot="People/save">Save</button>
   <button slot="People/cancel">Cancel</button>
```
```html
<!-- wrapped for styling, for example to make save button expand in width with the content -->
<div style="display: grid;...">
   <slot name="People/firstname"></slot>
   <slot name="Images/avatar"></slot>
   <!-- changed visual order -->
   <slot name="People/cancel"></slot>
   <slot name="People/save"></slot>
</div>

**Comment by warpech** [comment]
Regarding the example from https://github.com/Starcounter/Guidelines/issues/27#issuecomment-315014272, I think you meant that the `button`s are slots? You can still do this, just the inputs and the buttons are in different compositions:

```html
<!-- composition for: PersonPageForm.html (mapped to Person) -->
<slot name="People/firstname"></slot>
<slot name="Images/avatar"></slot>

<!-- composition for: PersonPageUI.html (not mapped) -->
<paper-button>Submit</paper-buton>
<paper-dialog>
   <slot name="People/save"></slot>
   <paper-buton>Cancel</paper-button>
</paper-dialog>
```

-----

Regarding the CSS Grid Layout example, you are right, I don't know how could we achieve such layout:

![artboard](https://user-images.githubusercontent.com/566463/28163042-b8af3c0a-67c9-11e7-8bf9-a4edeeac0a36.png)
 
The grid itself, Save and Cancel are defined in `PersonPageUI.html` (not mapped) and the labels and inputs are defined in `PersonPageForm.html` (mapped to Person).

Maybe with `display: subgrid` or `display: contents`?

Still I think that the benefits outweight the limitations.

**Comment by tomalec** [comment]
> I think you meant that the buttons are slots? 

I don't understand what you mean

 >  Maybe with `display: subgrid` or `display: contents`?

I would not expect any of those features to cross shadow dom boundary.

> Still I think that the benefits outweight the limitations.

What are the benefits of 5. given we already stated 1. and 2. ?(https://github.com/Starcounter/Guidelines/issues/27#issuecomment-309479099)

I would say:
>All our apps should have two levels of views:
> - lifecycle views (that have "Save", "Cancel", "Create new" buttons, nest blendable views)
> - blendable views (that do not have "Save", "Cancel", "Create new" buttons)

provides yet another complication to our (already hard to explain) blending feature, also it doubles the number of views, introduces confusing term "lifecycle view". Colides with our claim that "every view is blendable" (in basic degree).

**Comment by warpech** [comment]
>> I think you meant that the buttons are slots? 

> I don't understand what you mean

Disregard that, I didn't understand the example in the beginning. Then I forgot to remove this question.

> provides yet another complication to our (already hard to explain) blending feature, also it doubles the number of views, introduces confusing term "lifecycle view". Colides with our claim that "every view is blendable" (in basic degree).

"Lifecycle views" was not very informative term, but maybe "long-running transaction control views"?

> What are the benefits of 5. given we already stated 1. and 2. ?(#27 (comment))

The benefits are that:

- you don't have to do the duck-shooting for the buttons
- the host app provides the upper-level navigation and look and feel for Save and Cancel
- the initial composition today sucks by default 
    - after starting 3 apps usually you see: People inputs, People buttons, Images inputs, Images buttons, OtherApp inputs, OtherApp buttons.
    - you *must* make a custom composition to make it look reasonable
    - after point 5. is implemented, the default look would be: People inputs, Images inputs, OtherApp inputs, People buttons

Since we seem to not have agreement, let's keep what we have right now and implement only points 1-3 (https://github.com/Starcounter/Guidelines/issues/27#issuecomment-309479099). I will make a separate issue.

**Comment by tomalec** [comment]
>  maybe "long-running transaction control views"?

Waay more informative and self-explanatory to me :)

> -  you don't have to do the duck-shooting for the buttons

fair point :+1: 

> - the host app provides the upper-level navigation and look and feel for Save and Cancel

I would say it's questionable. To me it looks like a limitation so disadvantage.
We could use many other tools and techniques already (or almost already) available to serve desired look and feel from host app.

> - the initial composition today sucks by default

I would say it's exactly same point as the first one, just more elaborate.

On the other side no-5. benefits are:
- smaller Starcounter-specific vocabulary,
- more elements on partial level == more flexibility in custom design


> Since we seem to not have agreement, let's keep what we have right now and implement only points 1-3 (#27 (comment)). I will make a separate issue.

I'm all in for 1-3! 👍 
Given that 5. does not seem to require any specific features to be implemented - it's just a manner fo app architecture design, we can still play with it, try it here and there. But I will not force everybody to do that in docs.

Maybe, if we will have a case when 1-3 +is not enough and simple duck-shooting does not work or is problematic, we could continue the discussion. Then we may find some other ways to solve that problem.

**Comment by miyconst** [comment]
1. We should clearly distinguish what logic should and should not be in the save button.
1.1. Any database operations apart of transaction commit should not be included into the save button. We should keep in mind, that the database objects may be created through mapping, and the save button code will never be executed.
1.2. Any on-screen actions, such as form validation, can be executed within a save button handler.
2. The new version of `Db.Scope`, so called *Interaction Scope* is supposed to have `Committing` and `Committed` events. Kind of pre-commit and post-commit hooks, which work in scope of a single (~long-running~) transaction.
2.1. The `Committing` event is possible to cancel. 
2.2. Every app blended into the same scope will be able to subscribe to the `Committing` event.
2.3. The `Committing` event will be cancelled if one or more apps requested cancellation.
3. The pre-commit and post-commit hooks should be used for database updates and side effects.
3.1. *Send email* action should be implemented with a post-commit hook.
3.2. *Calculate computed properties cache* should be implemented with a pre-commit hook.

**Comment by warpech** [comment]
In my experience, it is not enough to establish such practice without enforcing it in the API. Can we consider enforcing that, for example by:
- disallow to explicitly call `Transaction.Commit()` and `Transaction.Cancel()` in `Input` method of TypedJSON
- instead use specialized bindings, such as:

```cs
public class PersonPage : Json
{
    static PersonPage()
    {
        DefaultTemplate.Save.Bind = Json.CommitCurrentTransaction;
        DefaultTemplate.Cancel.Bind = Json.RollbackCurrentTransaction;
    }
}
```

**Comment by miyconst** [comment]
>... disallow to explicitly call ...

I would not like to go this way. Let's say someone also wants to collect statistics of which button was clicked?

Let's discuss this offline later to the point.

**Comment by warpech** [comment]
> I would not like to go this way. Let's say someone also wants to collect statistics of which button was clicked?

He can still do that as a short-running transaction, he does not need a long-running transaction for that. This issue is about buttons that are bound to long-running transactions that are shared between apps.

**Comment by warpech** [comment]
It was discussed last week between @Starcounter-Jack, Rolf, @andwah and me that this will be solved by having a toolbar at the top of the page. This will emphasize that the Save and Cancel button affect the whole interaction scope, not just the current view.

The apps will provide `<input type="submit">` (and `<input type="reset">`?) to indicate that there is something Save'able and (Cancellable?). The apps can use callbacks `BeforeSave`, `AfterSave`, `BeforeCancel`, `AfterCancel` in TypedJSON.

Performing transaction commit and rollback, as well as presenting the toolbar, will be the responsibility of the host (TypedJSON and Starcounter app shell).

![customer save changes](https://user-images.githubusercontent.com/566463/41416636-a1490db8-6feb-11e8-9e3e-0f8b297308a9.png)

[Source of the image](https://drive.google.com/drive/u/0/folders/1JR_cErTrKIrwlgqRVYHsoX1u-uYjbyGy)

**Comment by andwah** [comment]
Some thoughts....

If the user has changed a partial and the "save toolbar" is showing and on the partial there is a link/button that opens up another partial, 

I guess one solution for this senario (if the user clicks on the link) is to block the user from navigating away from that partial (even if the user click "back" on the browser navigation).

**Comment by warpech** [comment]
> If the user has changed a partial and the "save toolbar" is showing and on the partial there is a link/button that opens up another partial,

According to @Starcounter-Jack this is not a problem but a feature. It allows you to look up some information in another app and then return to your work-in-progress.

But there are 2 challenges, obviously:

1. what if I don't know how to return (we might have URL for the page that made the session dirty)
2. how to highlight the changes that I made (he have meta information for that)

**Comment by warpech** [comment]
Throughout June we had discussions about this feature with @Starcounter-Jack and @rolfsmeds. I would like to get it finalized in the forthcoming weeks. The basic idea from the UX perspective is defined here: https://docs.google.com/document/d/1hlIlQK5J956WmtLDv_emCGWA2z5762z-wi047BebgBo/edit

TL;DR

- the app view does not render the Save and Cancel buttons
- the app view-model does not have Save and Cancel triggers nor trigger callbacks
- if the interactive scope (long running transaction) becomes dirty, Starcounter renders a global toolbar with Save and Cancel button
- clicking on this buttons commits/rolls back the transaction and calls lifecycle events in view-models `Commiting`, `Commited`, `Restoring`, `Restored`

Comments welcome (here or in the doc)

**Comment by bigwad** [comment]
>@bigwad is currently working on the new implementation of the Database Interaction Scope to replace the old and broken long running transaction. It should not be difficult to add those Commiting, Commited, Restoring, Restored events.
If delivery of the Database Interaction Scope moves far in the future, we will invest into extending existing Long Running Transactions with the required events.

As I see, there is no difference between predesigned long-running transactions](https://github.com/Starcounter/RebelsLounge/issues/372) and `interaction scope` except latter adds mapping consistency. In that regard, it's all ready to implement events you're talking about.

Btw, afaik `viewModel.Transaction` is outdated. One should use viewModel.AttachedScope instead, which doesn't expose `Transaction` object.

**Comment by miyconst** [comment]
>Btw, afaik viewModel.Transaction is outdated. One should use viewModel.AttachedScope instead, which doesn't expose Transaction object.

Then the only thing left is to extend the `viewModel.AttachedScope` object with the required events, right?

**Comment by LarsTjernstrom** [comment]
- *if the interactive scope (long running transaction) becomes dirty,
   Starcounter renders a global toolbar with Save and Cancel button*

And it should also check the new not implemented IsValid status. The
buttons should be visible if the scope becomes dirty, but not enabled if
there are validation errors.



[image: photo]

<https://www.facebook.com/StarcounterDB/>
<https://www.linkedin.com/company/starcounterdb>
<https://twitter.com/StarcounterDB>

 *Lars Tjernström *| *Software architect *
+46 70 422068003
<http://t.sidekickopen55.com/e1t/c/5/f18dQhb0S7lC8dDMPbW2n0x6l2B9nMJW7t5XZs8qStn2W2zW-gY64zKlYVQBb2Y56dStCf2-JMT-02?si=X-SIG-APPLE-NOKEY&pi=8F71530B-5836-4470-86D0-463AA94B2205&t=http%3A%2F%2Ft%2Esidekickopen55%2Ecom%2Fe1t%2Fc%2F5%2Ff18dQhb0S7lC8dDMPbW2n0x6l2B9nMJW7t5XZs8qStn2W2zW%2DgY64zKlYVQBb2Y56dStCf2%2DJMT%2D02%3Fsi%3DX%2DSIG%2DAPPLE%2DNOKEY%26pi%3D933EB8FF%2DBDA5%2D4644%2DB3FF%2D9117AF12413F%26t%3Dhttp%253A%252F%252Ft%252Esidekickopen54%252Ecom%252Fe1t%252Fc%252F5%252Ff18dQhb0S7lC8dDMPbW2n0x6l2B9nMJW7t5XZs8qStn2W2zW%252DgY64zKlYVQBb2Y56dStCf2%252DJMT%252D02%253Fsi%253DX%2DSIG%2DAPPLE%2DNOKEY%2526pi%253DE9813137%252DA95D%252D4994%252D9892%252D39A47DCF486F%2526t%253Dhttp%25253A%25252F%25252Ft%25252Esidekickopen51%25252Ecom%25252Fe1t%25252Fc%25252F5%25252Ff18dQhb0S7lC8dDMPbW2n0x6l2B9nMJW7t5XZs8qStn2W2zW%25252DgY64zKlYVQBb2Y56dStCf2%25252DJMT%25252D02%25253Fsi%25253DX%25252DSIG%25252DAPPLE%25252DNOKEY%252526pi%25253DE60A8C49%25252D695E%25252D4E4D%25252D9ED7%25252D2044EB4D773E%252526t%25253Dhttp%2525253A%2525252F%2525252Ft%2525252Esidekickopen42%2525252Ecom%2525252Fe1t%2525252Fc%2525252F5%2525252Ff18dQhb0S7lC8dDMPbW2n0x6l2B9nMJW7t5XZs8qStn2W2zW%2525252DgY64zKlYVQBb2Y56dStCf2%2525252DJMT%2525252D02%2525253Fsi%2525253DX%25252DSIG%25252DAPPLE%25252DNOKEY%25252526pi%2525253DAAB9D957%2525252DC78B%2525252D41D0%2525252DACB9%2525252D73F20BB47B41%25252526t%2525253Dhttp%252525253A%252525252F%252525252Ft%252525252Esidekickopen42%252525252Ecom%252525252Fe1t%252525252Fc%252525252F5%252525252Ff18dQhb0S7lC8dDMPbW2n0x6l2B9nMJW7t5XZs8qStn2W2zW%252525252DgY64zKlYVQBb2Y56dStCf2%252525252DJMT%252525252D02%252525253Fsi%252525253DX%252525252DSIG%252525252DAPPLE%252525252DNOKEY%2525252526pi%252525253DE894EC10%252525252DD5B5%252525252D44A6%252525252DA816%252525252D4D46AFBA1CD3%2525252526t%252525253Dhttp%25252525253A%25252525252F%25252525252Ft%25252525252Esidekickopen42%25252525252Ecom%25252525252Fe1t%25252525252Fc%25252525252F5%25252525252Ff18dQhb0S7lC8dDMPbW2n0x6l2B9nMJW7t5XZs8qStn2W2zW%25252525252DgY64zKlYVQBb2Y56dStCf2%25252525252DJMT%25252525252D02%25252525253Fsi%2525>
 | lars.tjernstrom@starcounter.com
www.starcounter.com | www.starcounter.io
Hovslagargatan 3, 111 48 Stockholm, Sweden

**Comment by miyconst** [comment]
>And it should also check the new not implemented IsValid status. The
buttons should be visible if the scope becomes dirty, but not enabled if
there are validation errors.

I would not go this way. Performing validation on every patch may be expensive.

P.S. @LarsTjernstrom please do not answer GitHub Emails.

**Comment by warpech** [comment]
> I would not go this way. Performing validation on every patch may be expensive.

I actually agree with @LarsTjernstrom. For a good UI, the save bar should now whether the submission is possible or not.

So I think we actually need a callback in TypedJSON to check whether a dirty view-model is committable or not.

Validation does not have to be performed on every patch. The view-model can cache its validity and return the cached value.

**Comment by miyconst** [comment]
>I actually agree with @LarsTjernstrom. For a good UX, the save bar must know whether the submission is possible or not.

I have a different opinion, but I am not a good UX designer. If the feature will not consume a lot of extra resources - I am fine.

**Comment by warpech** [comment]
As it was discussed today on the validation meeting, apart from the lifecycle events listed in https://github.com/Starcounter/Guidelines/issues/27#issuecomment-400636662, we will also need a `Validate` method and `Validating` event to support blocking the "Save" button for invalid user input.

More info: https://github.com/Starcounter/RebelsLounge/issues/275

**Comment by warpech** [comment]
Some thoughts about how the Save Bar could communicate with view-models to know if they are in dirty state: https://github.com/Starcounter/Minutes/issues/188#issuecomment-421329051

**Comment by warpech** [comment]
This discussion holds many ideas. The current idea that is the candidate for implementation starts from the comment https://github.com/Starcounter/Guidelines/issues/27#issuecomment-397305209 downwards.

The implementation is moved to another issue https://github.com/Starcounter/uniform/issues/203.

Archiving. Please continue the discussion in the new issue.
